modules/concurrency/include/hephaestus/concurrency/message_queue_consumer.h

```diff
@@ -58,6 +59,13 @@ void MessageQueueConsumer<T>::stop() {
   spinner_.stop();
 }
 
+template <typename T>
+void MessageQueueConsumer<T>::wait() {
+  while (!message_queue_.empty()) {
+    consume();
+  }
+}
```

Comment:
I think this logic should be moved inside the `stop` function:
```
template <typename T>
void MessageQueueConsumer<T>::stop() {
  message_queue_.stop(); // We do not accept messages anymore
  spinner_.stop(); // We make sure that the spinner is no longer accessing the queue
 // We go through all the messages left and we consume them
  while (!message_queue_.empty()) {
    consume();
  }
}
```
At this point to make it perfect I would also probably make `stop` return a `std::future` so that we do not block while stopping.

Also I would remove the `wait` queue all together from the public interface as it can break the spinner or the ordering.

Reply:
great, much better design!